### PR TITLE
Update Object3D.updateMatrixWorld() to update parent matrixworlds

### PR DESCRIPTION
**Pull request by gero3**
This is a better solution for #6426

Things to note:
- It doesn't check if everything is already updated.
- It uses recursion, so is slower then it could be

### CODE REVIEW

**Comment by WestLangley** [comment]
@gero3 Your new method  `getUpdatedMatrixWorld()` is quite clever, IMHO, and is a solution to a much larger issue.

`getUpdatedMatrixWorld()` traverses up a single branch of the tree, and ignores the children of the object -- as it should. The method `updateMatrixWorld()`, on the other hand, traverses down the tree and updates the object's children.

If I am not mistaken, what we should do is (1) rename `updateMatrixWorld()` to something else, like `updateWorldMatrices()`, and (2) rename your new method to  `updateMatrixWorld()`.

If we did that, `scene.updateWorldMatrices()` would then be called by the renderer to traverse the scene graph, and your new version of `updateMatrixWorld()` would do what it should have been doing all along.

This will have no effect on users. Developers can call `updateMatrixWorld()` prior to referencing `object.MatrixWorld` in cases where the traversal would not be too burdensome.

**Comment by mrdoob** [comment]
Wow. Mind blown. So pretty :+1: 
Also, what @WestLangley said!

**Comment by gero3** [comment]
These are the places that needs to be checked if we rename.

```
/src/cameras/CubeCamera.js:
	50: 		if ( this.parent === undefined ) this.updateMatrixWorld();

/src/core/Object3D.js:
	414: 		this.updateMatrixWorld( true );
	429: 			this.updateMatrixWorld( true );
	464: 			this.updateMatrixWorld( true );
	538: 	updateMatrixWorld: function ( force ) {
	564: 			this.children[ i ].updateMatrixWorld( force );

/src/extras/audio/Audio.js:
	122: THREE.Audio.prototype.updateMatrixWorld = ( function () {
	128: 		THREE.Object3D.prototype.updateMatrixWorld.call( this, force );

/src/extras/audio/AudioListener.js:
	18: THREE.AudioListener.prototype.updateMatrixWorld = ( function () {
	28: 		THREE.Object3D.prototype.updateMatrixWorld.call( this, force );

/src/extras/core/Gyroscope.js:
	14: THREE.Gyroscope.prototype.updateMatrixWorld = ( function () {
	59: 			this.children[ i ].updateMatrixWorld( force );

/src/extras/helpers/DirectionalLightHelper.js:
	12: 	this.light.updateMatrixWorld();

/src/extras/helpers/FaceNormalsHelper.js:
	48: 	object.updateMatrixWorld( true );

/src/extras/helpers/HemisphereLightHelper.js:
	11: 	this.light.updateMatrixWorld();

/src/extras/helpers/PointLightHelper.js:
	9: 	this.light.updateMatrixWorld();

/src/extras/helpers/SpotLightHelper.js:
	12: 	this.light.updateMatrixWorld();

/src/extras/helpers/VertexNormalsHelper.js:
	53: 		this.object.updateMatrixWorld( true );

/src/extras/helpers/VertexTangentsHelper.js:
	52: 		this.object.updateMatrixWorld( true );

/src/math/Box3.js:
	68: 			object.updateMatrixWorld( true );

/src/objects/SkinnedMesh.js:
	75: 	this.updateMatrixWorld( true );
	90: 		this.updateMatrixWorld( true );
	137: THREE.SkinnedMesh.prototype.updateMatrixWorld = function( force ) {
	139: 	THREE.Mesh.prototype.updateMatrixWorld.call( this, true );

/src/renderers/WebGLRenderer.js:
	1634: 		if ( scene.autoUpdate === true ) scene.updateMatrixWorld();
	1638: 		if ( camera.parent === undefined ) camera.updateMatrixWorld();

/src/renderers/webgl/WebGLShadowMap.js:
	204: 				if ( scene.autoUpdate === true ) scene.updateMatrixWorld();
	230: 			shadowCamera.updateMatrixWorld();

Found 28 matches in 16 files 
```

**Comment by mrdoob** [comment]
Wait. Thinking about this a bit more...

This can break a ton of code out there. Anyone extending `Object3D` and doing their own `updateMatrixWorld()`...

Sorry about this @gero3 but I think it's better to revert to the original commit. Naming-wise `.getUpdatedMatrixWorld()` is probably ok, but how about `.traverseUpdateMatrixWorld()` or `.parentUpdateMatrixWorld()`?

**Comment by WestLangley** [comment]
@mrdoob I am not following you. The current version of `updateMatrixWord()` does not "update" `matrixWorld` because it does not update parents. Instead, it does something else. @gero3's new method does what `updateMatrixWord()` should be doing. We should fix it.

**Comment by mrdoob** [comment]
@WestLangley Yes, I agree with that. The problem is that we should be careful on not breaking other people's code. With this change we're changing what `.updateMatrixWorld()` does (no longer updates the children, so this will be breaking the code of people that depend on this behaviour.

**Comment by WestLangley** [comment]
@mrdoob  I hear you, but we are compelled to fix things that are not correct. I believe the benefits of this change outweigh the costs -- unless a performance study indicates otherwise.

**Comment by mrdoob** [comment]
The problem is that by just changing the behaviour of a method we can't let the user know. One option would be to find new names for the methods and redirect with a warning.

Something like `.getUpdatedMatrixWorld()` and `.computeWorldMatrices()`. That way, if someone extends `updateMatrixWorld()` we can warn the user that they its been renamed to `.computeWorldMatrices()`.

**Comment by WestLangley** [comment]
OK. To cover all the bases, I think we only need the following two methods:

`object.updateMatrix()` - updates `object.matrix` from `object.position`, `object.quaternion`, and `object.scale`. This method does not change.

`object.updateWorldMatrix( recursive, force )` - updates `object.matrixWorld` using the approach proposed by @gero3. If `recursive` is true (it defaults to `false`), it also updates the world matrices of the object's children. (The `force` flag is something we may need -- may not... )

So, the renderer would call `scene.updateWorldMatrix( true )`. Users would be advised to use `updateWorldMatrix()` instead of `updateMatrixWorld()`.

---

Alternatively, we can keep things really simple, and keep the `updateMatrixWorld()` method -- just modify it so it traverses up the tree, too.

**Comment by RemusMar** [comment]
If the (wrong) backward compatibility is so important to Ricardo,
I've already gave him an idea: add an extra parameter for ADD and DETACH.
Something like:
group.add( child, preserve);
Where "preserve" is true or false (false by default).

And something more for "gero3":
Simply by adding a updateMatrixWorld() before THREE.SceneUtils.detach, it works.
https://jsfiddle.net/46n9rjw6/2/ 

That's a partial fix.
As you can see, after "cube.add( cube2 );", the new "cube" parent messed up the size (scale) of the new "cube2" child.
cube2 is now half size compared to its initial size.
And that's wrong!

**Comment by gero3** [comment]
@remusMar That is what we are fixing now

**Comment by WestLangley** [comment]
I still think what I proposed earlier is a good idea:

> Alternatively, we can keep things really simple, and keep the updateMatrixWorld() method -- just modify it so it traverses up the tree, too.

So `updateMatrixWorld()` would be modified to perform whatever calculations are required to compute the correct world matrix for the object -- and its children.

@gero3 has provided a clever, recursive solution that can be modified to do just that. If there are no performance issues, can we agree on a course of action?

File: src/core/Object3D.js [hunk 1]
```diff
@@ -561,11 +561,29 @@ THREE.Object3D.prototype = {
...
+
+		if ( this.matrixAutoUpdate === true ) this.updateMatrix();
+
+		if ( this.parent === undefined ) {
+
+			this.matrixWorld.copy( this.matrix );
+
+		} else {
+
+			this.matrixWorld.multiplyMatrices( this.parent.getUpdatedMatrixWorld() , this.matrix );
```
Comment (mrdoob) [inline src/core/Object3D.js]: Shouldn't this be ```this.parent.updateMatrixWorld()```?

**Comment by mrdoob** [comment]
Naming wise, and for clarity... How about we keep `updateMatrixWorld` as it is and we name the new method as `solveMatrixWorld`, `traverseMatrixWorld`, `computeMatrixWorld`, ...?

**Comment by WestLangley** [comment]
> Naming wise, and for clarity... How about we keep updateMatrixWorld as it is and we name the new method as solveMatrixWorld, traverseMatrixWorld, computeMatrixWorld, ...?

I think that is a mistake. The current method, `updateMatrixWorld()`, does not correctly update the world matrix. It does not do what its name says it does. It needs to be modified so it updates the world matrices of the object's parent(s) first.

We do not need any new methods. We just need to fix the method we have. That is the simple, backwards-compatible, solution.

**Comment by RemusMar** [comment]
> We do not need any new methods. We just need to fix the method we have.
> That is the simple, backwards-compatible, solution.

+1

**Comment by gero3** [comment]
I tried to traverse to update childs.
This code hasn't been tested yet

**Comment by WestLangley** [comment]
Signature has to change?

     updateMatrixWorld( recursive, noUpdatingParent )

I think we need to find a solution where the signature to this public method remains the same... although I always found the `force` flag somewhat confusing. I'd love to remove that arg. Thinking...

BTW, can you squash these commits for ease of reading?

**Comment by mrdoob** [comment]
> We do not need any new methods. We just need to fix the method we have. That is the simple, backwards-compatible, solution.

I don't think that's possible.

Before doing a render we have to traverse update all the `matrixWorld`s. With the proposed method, how would we update all the `matrixWorld`s without updating the parents one many times?

Say we have this hierarchy:

```
Scene
    Group
        Mesh
        Mesh2
```

The current [updateMatrixWorld](https://github.com/mrdoob/three.js/blob/master/src/core/Object3D.js#L538-L568) does this:

```
1. Update Scene, then updates children:
    2. Update Group, then updates children:
        3. Update Mesh and Mesh2
```

The proposed method would be doing this:

```
1. Update Scene, then update children:
    2. Update Scene, then update Group, then update children:
        3. Update Scene, then update Group, then update Mesh, then update Scene, then update Group, then update Mesh2
```

Am I missing something?

**Comment by mrdoob** [comment]
Oh, I see now that @gero3 just added `noUpdatingParent` to the method to solve this...

**Comment by mrdoob** [comment]
How about `updateMatrixWorld` behaves as the new proposed method, and we rename the old one to `traverseUpdateMatrixWorld`?

**Comment by WestLangley** [comment]
I am concerned about the signature change. How about the following, with all args optional?

    updateMatrixWorld( force, recursive, noUpdatingParent )

> rename the old one to `traverseUpdateMatrixWorld`?

Do we need more than one method?

**Comment by mrdoob** [comment]
It's just that this is yet another if statement in a pretty hot code. It may not make much difference though.

**Comment by gero3** [comment]
This makes the current version work with all webgl examples (tested) except for ParallaxBarrierEffect.

**Comment by mrdoob** [comment]
Hmm... What about this...?

1. We add `this._matrix = new THREE.Matrix4()` and `this._matrixWorld = new THREE.Matrix4()` to `Object3D`.
2. `this.matrix` turns into a getter:

```javascript
get matrix () {
	return this.updateMatrix();
};
```

3. `this.matrixWorld` turns into a getter:

```javascript
get matrixWorld () {

	if ( this.matrixAutoUpdate === true ) this.updateMatrix();

	if ( this.parent === undefined ) {

		this._matrixWorld.copy( this._matrix );

	} else {

		this._matrixWorld.multiplyMatrices( this.parent.matrixWorld , this._matrix );

	}
		
	return this._matrixWorld;

};
```

4. We replace all the relevant calls to `.matrix` and `.matrixWorld` to `._matrix` and `._matrixWorld` for performance.

This way the user will always have the right matrix when accessing `.matrix` and `.matrixWorld`, we keep the same API and we avoid breakage.

**Comment by gero3** [comment]
few things to note: 

1. Wouldn't MatrixCached and  matrixWorldCached be a better name?? It would also make a lot more clear that they don't get updated a la minute. 

2. This changes the performance charateristics so this should be rather well documented.

**Comment by gero3** [comment]
#6851 is a rapid try at the new version

**Comment by gero3** [comment]
Closing this one since the other one seems better

**Pull request closed without merging**